projects/plugins/boost/app/modules/cache/Boost_File_Cache.php

```diff
@@ -0,0 +1,113 @@
+	 * @return string - The path to the cache directory for the given request.
+	 */
+	private function path( $request_uri = false ) {
+		if ( $request_uri !== false ) {
+			$request_uri = $this->normalize_request_uri( $request_uri );
+		} else {
+			$request_uri = $this->request_uri;
+		}
+
+		$key  = $this->path_key( $request_uri );
+		$path = WP_CONTENT_DIR . '/boost-cache/cache/';
+
+		/*
+		 * The cache directory is split into 5 levels of subdirectories.
+		 * 2 characters of the md5 hash of the request uri are used for each level.
+		 * This is done to prevent having too many files in a single directory.
+		 */
+		for ( $i = 0; $i < 10; $i += 2 ) {
+			$path .= substr( $key, $i, 2 ) . '/';
+		}
+
+		return $path;
+	}
+
+	/*
+	 * Returns the cache filename for the given request, or current request.
+	 *
+	 * @param array $args - an array containing the request_uri, cookies array, and get array representing the request.
+	 * @return string - The cache path + filename for the given request.
+	 */
+	private function cache_filename( $args = array() ) {
+		$defaults = array(
+			'request_uri' => $this->request_uri,
+			'cookies'     => $_COOKIE, // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+			'get'         => $_GET, // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized, WordPress.Security.NonceVerification.Recommended
+		);
+		$args     = array_merge( $defaults, $args );
+
+		return $this->path( $args['request_uri'] ) . $this->cache_key( $args ) . '.html';
+	}
+
+	/*
+	 * Outputs the cached page if it exists for the given request, or current request.
+	 *
+	 * @param array $args - an array containing the request_uri, cookies array, and get array representing the request.
+	 * @return bool - false if page was not cached.
+	 */
+	public function get( $args = array() ) {
+		if ( ! $this->is_cacheable() ) {
+			return false;
+		}
+		if ( file_exists( $this->cache_filename( $args ) ) ) {
+			// phpcs:ignore WordPress.WP.AlternativeFunctions.file_get_contents_file_get_contents, WordPress.Security.EscapeOutput.OutputNotEscaped
+			echo file_get_contents( $this->cache_filename( $args ) ) . '<!-- cached -->';
+			die();
+		}
+		return false;
+	}
+
+	/*
+	 * Creates the cache directory if it doesn't exist.
+	 *
+	 * @param string $path - The path to the cache directory to create.
+	 */
+	private function create_cache_directory( $path ) {
+		if ( ! is_dir( $path ) ) {
+			// phpcs:ignore WordPress.WP.AlternativeFunctions.dir_mkdir_dirname, WordPress.WP.AlternativeFunctions.file_system_operations_mkdir
+			return mkdir( $path, 0755, true );
+		}
+
+		return true;
+	}
+
+	/*
+	 * Saves the output buffer to the cache file for the given request, or current request.
+	 * Then outputs the buffer to the browser.
+	 *
+	 * @param string $data - The output buffer to save to the cache file.
+	 * @return bool - false if page was not cacheable.
+	 */
+	public function set( $data ) {
```

Comment:
Just a minor suggestion.
IMHO, it returning the cached data can cause unintentional bugs. If the call depends on this for output and we end up updating this function to return false for more cases than `strlen( $data ) === 0`, we may end up in situations where we are returning blank pages where we could send the data.

In short, I think this should return boolean true or false. The caller would use this only to set the cache, but output it independently.

Reply:
Hmm, I see what you mean. The set function is used as the callback for ob_start(), but maybe the callback should be "ob_callback" and it should call set() to record the output buffer?

The output buffer callback returns the $data so that PHP sends it to the browser, but you're right. set() should return a bool to signal that there was a problem writing the cache file.